### PR TITLE
Drop titles repeating within 300 seconds

### LINKED ISSUE #183: skip repeating url titles so often
how about a little cache [(url, timestamp)], say with max 5 members, fifo, for begiak's url title thing? That way we might skip repeating titles said within last N secs.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #183
@shardulc Friendly ping.

### CODE REVIEW

File: modules/head.py [hunk 1]
```diff
@@ -92,6 +96,17 @@ def noteuri(phenny, input):
 
 def snarfuri(phenny, input):
     uri = input.group(2)
+
+    nowtime = time.time()
+
+    if uri in phenny.recent_titles:
+        oldtime = phenny.recent_titles[uri]
+
+        if nowtime - oldtime < 300:
```
Comment (sushain97) [inline modules/head.py]: Ideally magic numbers should be defined in a constant at the top of the file

Comment (Androbin) [reply modules/head.py]: Sure thing, what would be a good name?

Comment (Androbin) [reply modules/head.py]: Something like `title_max_repeat_time` ?

Comment (sushain97) [reply modules/head.py]: Sure, as long as it goes with a comment e.g. `# seconds until a URL title will be repeated if said multiple times`

**Comment by shardulc** [comment]
@Androbin As before, could you please provide some logs/screenshots of the new behavior?

**Comment by Androbin** [comment]
screenshot

**Comment by shardulc** [comment]
What about Wikipedia links? I believe the `wikipedia.py` also has some posted db stuff?

**Comment by Androbin** [comment]
> What about Wikipedia links? I believe the wikipedia.py also has some posted db stuff?

Wikipedia links are handled by `head.gettitle` but `posted.check_posted` is not called for them directly, but instead `wikipedia.wikipedia` does this. The only deal here is that the Wikipedia module does some minor wrapping around `posted.check_posted`.

**Comment by shardulc** [comment]
@Androbin Cool. Last thing, add a test method to `test_head.py` to test this. Don't actually wait 300 seconds in the test, instead:
 1. invoke `snarfuri` and make sure a time is added to `recent_titles`
 1. invoke `snarfuri` again and make sure it doesn't say anything
 1. change the time in `recent_titles` manually by 300 seconds
 1. invoke `snarfuri` again and make sure it says something

**Comment by sushain97** [comment]
There are also simple ways to mock time that would be a bit more ideal, IMO.

**Comment by Androbin** [comment]
@sushain97 Thought of that, too. But I'd rather not depend on `snarfuri` using a particular method.

**Comment by sushain97** [comment]
It seems to be far better than relying exactly on how `snafuri` works internally, i.e. `recent_titles`.

**Comment by Androbin** [comment]
@sushain97 Do you propose providing an interface for the manipulation of i.e. `recent_titles`?

**Comment by shardulc** [comment]
@Androbin I think he means you should mock the `time.time()` method.

**Comment by Androbin** [comment]
@shardulc I think he means better not knowing that `snarfuri` uses `time.time()` internally.

**Comment by sushain97** [comment]
> @sushain97 Do you propose providing an interface for the manipulation of i.e. recent_titles?

Really overkill :)

> @Androbin I think he means you should mock the time.time() method.

This. Sure, `snafuri` could conceivably use something else but that would probably be equally easy to mock.

**Comment by shardulc** [comment]
Looks good. @Androbin please remove the database file from version control. Oddly, tests are failing due to the weather module again...

**Comment by wei2912** [comment]
Have the recent changes been tested?

**Comment by shardulc** [comment]
@wei2912 Yup, plus tests were written for most of them.

**Comment by Androbin** [comment]
@shardulc We're good to go.

**Comment by Androbin** [comment]
@sushain97 We're good to go.

**Pull request merged**